### PR TITLE
Add MapDataset.stack

### PR DESCRIPTION
**Pull request by AtreyeeS**
This is a preliminary implementation for `MapDataset.stack`. It does not check if `other` is a cutout of `self`, but implicitly assumes so.

### CODE REVIEW

File: gammapy/cube/fit.py [hunk 1]
```diff
@@ -346,6 +346,49 @@ def create(
...
+
+        if isinstance(other.edisp, EDispMap):
+            self.edisp.stack(other.stack())
+        elif isinstance(self.psf, EnergyDispersion):
+            raise ValueError("Stacking of edisp kernels not supported")
+
+        if self.gti:
+            self.gti = self.gti.stack(other.gti).union()
+
+        # How to stack the mask?
```
Comment (AtreyeeS) [inline gammapy/cube/fit.py]: We should typically want to do logical `or` on the mask as in `SpectrumDataset.stack()`.
However, `mask_fit` is not a `Map` object but a numpy array. How do we take care of the fact that `other.mask_fit` is a cutout of `self.mask_fit` and will not have the same dimensions?

File: gammapy/cube/fit.py [hunk 2]
```diff
@@ -346,6 +346,49 @@ def create(
...
+        """
+        Stack the maps in other onto self.
+        The stacking is done inplace
+
+        Parameters
+        ----------
+        other: `~gammapy.cube.MapDataset`
+            the MapDatasets to be stacked with this one.
+
+        Returns
```
Comment (adonath) [inline gammapy/cube/fit.py]: As the stacking is done in place, nothing is returned. So just remove this section...

File: gammapy/cube/fit.py (hunk 1 above)
Comment (adonath) [reply gammapy/cube/fit.py]: This is a good point and probably comes in as an argument to make the `mask_fit` and `mask_safe` a `Map` object as well. For now I think it's OK to temporarily create a map object and use it to stack the mask like so:
```python3

mask_safe = Map.from_geom(self.counts.geom, data=self.mask_safe)
mask_safe_other = Map.from_geom(other.counts.geom, data=other.mask_safe)
```
Note that the `+` operator on `bool` types should already give what we want:
```python3
import numpy as np
a = np.array([1, 1, 0, 0], dtype=bool)
b = np.array([1, 0, 1, 0], dtype=bool)
a + b
```
Gives:
```
array([ True,  True,  True, False])
```

File: gammapy/cube/fit.py [hunk 3]
```diff
@@ -346,6 +346,49 @@ def create(
...
+        -------
+        new : `~gammapy.cube.MapDataset`
+            A dataset with the stacked maps
+
+        """
+
+        if self.counts:
+            if other.mask_fit:
+                other.counts = other.counts * other.mask_fit
```
Comment (adonath) [inline gammapy/cube/fit.py]: I think we should not modify the `other` object when stacking. For the "in place" stacking we should only modify `self`. My proposal would be to add a `weights` argument (just as a numpy array) to `Map.coadd()`, which is multiplied to the values stacked in the map. So the code here would be:
```python3
self.counts.coadd(other.counts, weights=other.mask_safe)
```
Also note that we want to use `mask_safe` for the stacking not `mask_fit`.

File: gammapy/cube/fit.py [hunk 4]
```diff
@@ -346,6 +346,49 @@ def create(
...
+            self.counts.coadd(other.counts)
+
+        self.exposure.coadd(other.exposure)
+
+        if other.mask_fit:
+            other.background_model.map = other.background_model.map * other.mask_fit
+        self.background_model.map.coadd(other.background_model.evaluate())
```
Comment (adonath) [inline gammapy/cube/fit.py]: Same comment as above, maybe we should add a `weights` argument to `Map.coadd()`.

**Comment by AtreyeeS** [comment]
Thanks @adonath ! Addressed the changes. For now, I am going with coadd, and we can shift to a faster implementation (add a `Map.stack()` or improve `Map.coadd()` for aligned geoms as noted in the TODO

**Comment by adonath** [comment]
Thanks a lot @AtreyeeS!

**Pull request merged**